### PR TITLE
[MPS] Migrate nll_loss to native Metal kernels

### PR DESCRIPTION
**Pull request by anagnorisis2peripeteia**
# [MPS] Migrate nll_loss to native Metal kernels

Part of the MPSGraph shape-keyed graph-cache elimination effort (#187455). Replaces the MPSGraph implementation of 1-D `nll_loss` forward and backward with Metal kernels (two-phase float-accumulated reduction for mean/sum), removing the 1-D nll graph-cache sites. `nll_loss2d` stays on MPSGraph as a follow-up. Rebased over trunk's new ctc_loss kernels in the same files.

## Safety and correctness

- **Targets and `ignore_index` ride at int64 end-to-end** — no narrowing. Previously-drafted int32 handling silently wrapped `ignore_index=2^32` to 0 (changing which classes are ignored) and indexed wrong classes for out-of-int32 targets. Also removes the target conversion copy on the hot path (`to(kLong).contiguous()` are both no-ops for dense Long targets).
- Three kernels validate `target ∈ [0, C)` via the async error buffer and raise on out-of-bounds like CPU; `ignore_index` outside `[0, C)` never raises and ignores correctly.
- **All-zero-weight mean backward matches CPU's NaN gradients** (0/0 propagates; no denominator fallback).
- Inputs above 2^32 elements are rejected explicitly (kernels index 32-bit).
- Partial sums accumulate in float for half/bfloat; f16 large-N accuracy exceeds the CPU-f16 kernel. Reduction is bitwise deterministic (fixed two-phase pass, no atomics).
- **Fixes a pre-existing MPSGraph failure**: f16 inputs with N > 65504 threw "value cannot be converted to type c10::Half without overflow" on the graph path; the Metal path matches CPU-f16 semantics instead.

Known divergence (nit): C == 0 with non-empty non-ignored targets returns zeros/NaN instead of CPU's IndexError (validating would need a host-side target read).

## Perf

`benchmarks/mps/bench_nll_loss.py` (Timer.blocked_autorange, K=20 amortized behind one synchronize, median of 3 interleaved A/B reps vs stock nightly), M4 Pro — 29/29 comparable cells, 0 regressions:

| cell | dtype | shape | speedup |
|---|---|---|---|
| fwd_mean | f32 | N4096×C10 | 3.70x |
| fwd_mean | f32 | N128×C32000 | 3.13x |
| fwd_mean | f32 | N4096×C1000 | 2.39x |
| fwd_mean | f32 | N65536×C10 | 2.34x |
| fwd_none | f32 | N4096×C1000 | 3.64x |
| fwd_mean_w | f32 | N4096×C10 | 4.25x |
| bwd_mean | f32 | N4096×C1000 | 3.20x |
| bwd_mean | f32 | N65536×C10 | 1.19x |
| fwd_mean | f16 | N128×C32000 | 3.20x |
| bwd_mean | f16 | N4096×C1000 | 2.47x |
| (remaining 19 cells) | | | 1.19–4.14x |

The four f16 N=65536 cells have no baseline column: the MPSGraph path hard-errors on them (the overflow bug above); the Metal path runs them at f32-comparable speed.

## Not tested

- `nll_loss2d` (still MPSGraph; deliberate follow-up — the spatial bench cell documents the boundary).
- complex (not supported by nll_loss on any backend).

## Test plan

```
python -m pytest test/test_mps.py -k nll
python -m pytest test/test_ops.py -k "mps and nll"
python -m pytest test/test_modules.py -k "mps and NLL"
python benchmarks/mps/bench_nll_loss.py
```

Full `--mps` CI matrix run locally (13 files; only failure is the known `test_metal` desktop-build artifact). Plus a 49-check standalone review-repro suite committed in the PR worktree covering OOB/ignore_index edges, weights (zero/negative/noncontiguous/dtype-mismatch), backward for all reductions with ignored positions, noncontiguous self/target/grad_output/grad_input, stride-0 grad_output, reduction determinism, and f16/bf16 accumulation accuracy.

Review log: https://gist.github.com/anagnorisis2peripeteia/f8cf5beaab478a440eb86a6a08aca74f

### CODE REVIEW

File: aten/src/ATen/native/mps/kernels/LossOps.metal [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+  threadgroup_barrier(mem_flags::mem_threadgroup);
+  if (lid == 0u) {
+    float acc = 0.f;
+    for (uint g = 0u; g < (tgsz >> 5u); ++g)
+      acc += float(smem[g]);
+    smem[0] = T(acc);
+  }
+  threadgroup_barrier(mem_flags::mem_threadgroup);
+  return smem[0];
+}
```
Comment (malfet) [inline aten/src/ATen/native/mps/kernels/LossOps.metal]: There is already one in reduce.h

File: aten/src/ATen/native/mps/kernels/LossOps.metal [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+// Param structs  (binary layout must match C++ structs in LossOps.mm)
+// ──────────────────────────────────────────────────────────────────────────
+
+struct NLLParams {
+  uint32_t N;
+  uint32_t C;
+  int32_t ignore_index;
+  uint32_t reduction;
+  uint32_t has_weight;
+};
```
Comment (malfet) [inline aten/src/ATen/native/mps/kernels/LossOps.metal]: Should be in the header

**Comment by malfet** [comment]
@claude review this and pay attention to code duplication

**Comment by malfet** [comment]
@claude review this please

File: benchmarks/mps/bench_nll_loss.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+"""
+Official PyTorch blocked_autorange benchmark for NLL loss on MPS.
```
Comment (malfet) [inline benchmarks/mps/bench_nll_loss.py]: Why benchmark is specific to MPS?

File: benchmarks/mps/bench_nll_loss.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+from torch.utils.benchmark import Timer
+
+DEVICE = "mps"
+_ = torch.randn(1024, device=DEVICE).sum()
+torch.mps.synchronize()
+
+DTYPES = [torch.float32, torch.float16, torch.bfloat16]
+REDUCTIONS = ["none", "mean", "sum"]
+# N=batch_size, C=num_classes
+CONFIGS = [(128, 1000), (1024, 100), (65536, 10), (1024, 10000), (8192, 4096)]
```
Comment (malfet) [inline benchmarks/mps/bench_nll_loss.py]: Why all shapes are even and for the most part powers of two? Would be nice to have them tied to some common model shapes

File: benchmarks/mps/bench_nll_loss.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    print(
+        f"  {str(cfg):<18} {str(dtype).split('.')[-1]:<10} {red:<8}"
+        f" {m.mean * 1e6:>10.2f} {m.median * 1e6:>10.2f}"
+    )
+
+
+hdr("NLLLoss (blocked_autorange)")
+for cfg, dtype, red in itertools.product(CONFIGS, DTYPES, REDUCTIONS):
+    N, C = cfg
+    lp = F.log_softmax(torch.randn(N, C, dtype=dtype, device=DEVICE), dim=1)
```
Comment (malfet) [inline benchmarks/mps/bench_nll_loss.py]: How about non-contiguous inputs?

File: aten/src/ATen/native/mps/kernels/LossOps.metal (hunk 1 above)
Comment (anagnorisis2peripeteia) [reply aten/src/ATen/native/mps/kernels/LossOps.metal]: Dropped the hand-rolled `tg_sum` and switched both reduce kernels to `c10::metal::threadgroup_sum<float>` from `reduction_utils.h`. Accumulation stays in float as before.

File: aten/src/ATen/native/mps/kernels/LossOps.metal (hunk 2 above)
Comment (anagnorisis2peripeteia) [reply aten/src/ATen/native/mps/kernels/LossOps.metal]: Moved `NLLParams` into a new `kernels/LossOps.h` (mirroring `EmbeddingBag.h`), now included from both the `.metal` and the `.mm`.

File: benchmarks/mps/bench_nll_loss.py (hunk 3 above)
Comment (anagnorisis2peripeteia) [reply benchmarks/mps/bench_nll_loss.py]: Added a `--device` flag and a generic `synchronize()`; the script now runs on CUDA/CPU too, with MPS as the default.

File: benchmarks/mps/bench_nll_loss.py (hunk 4 above)
Comment (anagnorisis2peripeteia) [reply benchmarks/mps/bench_nll_loss.py]: Replaced the power-of-two shapes with real classifier/LM-head sizes: ImageNet-1k/22k heads and Llama-2/3 + GPT-2 vocab sizes.

File: benchmarks/mps/bench_nll_loss.py (hunk 5 above)
Comment (anagnorisis2peripeteia) [reply benchmarks/mps/bench_nll_loss.py]: Added a non-contiguous layout case (a transposed `(C, N)` tensor) so the benchmark also covers the `.contiguous()` fallback in the dispatch.

**Comment by anagnorisis2peripeteia** [comment]
Thanks for the review. Addressed all of the above in `fc26ad2`:
- `tg_sum` removed in favor of `c10::metal::threadgroup_sum` from `reduction_utils.h`.
- `NLLParams` moved to a shared `kernels/LossOps.h`.
- Benchmark is now device-generic, uses real model-head shapes, and covers non-contiguous inputs.

Also took the bot's correctness notes: `total_weight` for the `None` reduction now matches the ATen reference (0 rather than N), and the forward/backward Mean-denominator asymmetry is documented in a comment.

**Comment by anagnorisis2peripeteia** [comment]
Closing as partially scooped.

Isalia20 landed native Metal `nll_loss` backward in #195153 (`754c3c9b352f` on `main`), with a follow-up 1-D fix in #195411.

This PR still mixed forward+backward. Rebasing it would fight that landed backward kernel.

`nll_loss` forward is still MPSGraph (`nll_loss_forward_out_mps` in `LossOps.mm`). That remaining site stays on #187455 as unclaimed follow-up, not as a rebase of this PR.
